monty-31/src/aarch64_neon/poseidon2.rs

```diff
@@ -49,28 +143,181 @@ impl<FP: FieldParameters, const WIDTH: usize, ILP: InternalLayerBaseParameters<F
+                ILP::diagonal_mul(&mut internal_state.s_hi);
+
+                // Compute the total sum of the entire state (`s0` + `sum_tail`)
+                // after the S-box on `s0` is complete.
+                let sum = sum_tail + internal_state.s0;
+
+                // Update `s0`. The transformation is `s0_new = sum + v_0*s0`.
+                //
+                // Since the diagonal element `v_0` is always -2, this becomes `sum - 2*s0`.
+                //
+                // For efficiency, this is calculated as `(sum - s0) - s0`, which is `sum_tail - s0`.
+                internal_state.s0 = sum_tail - internal_state.s0;
+
+                // Update the rest of the state. For each `s_i` (where `i > 0`), the
+                // transformation is `s_i_new = (v_i * s_i) + sum`. The `v_i * s_i` part
+                // was already computed in `diagonal_mul`.
+                ILP::add_sum(
+                    &mut internal_state.s_hi,
+                    transmute::<PackedMontyField31Neon<FP>, uint32x4_t>(sum),
+                );
+            });
+
+            // Convert the specialized state representation back to a standard array before returning.
+            *state = InternalLayer24::to_packed_field_array(internal_state);
+        }
+    }
+}
+
+/// Represents the external layers (initial and terminal) of the Poseidon2 permutation for NEON-accelerated operations.
+#[derive(Debug, Clone)]
+pub struct Poseidon2ExternalLayerMonty31<MP: MontyParameters, const WIDTH: usize> {
+    /// The original, scalar round constants for both initial and terminal external rounds.
+    pub(crate) external_constants: ExternalLayerConstants<MontyField31<MP>, WIDTH>,
+    /// Pre-processed constants for the initial external rounds, packed into NEON vectors in negative form (`c - P`).
+    packed_initial_external_constants: Vec<[uint32x4_t; WIDTH]>,
+    /// Pre-processed constants for the terminal external rounds, packed into NEON vectors in negative form (`c - P`).
+    packed_terminal_external_constants: Vec<[uint32x4_t; WIDTH]>,
+}
+
+impl<FP: FieldParameters, const WIDTH: usize> ExternalLayerConstructor<MontyField31<FP>, WIDTH>
+    for Poseidon2ExternalLayerMonty31<FP, WIDTH>
+{
+    fn new_from_constants(
+        external_constants: ExternalLayerConstants<MontyField31<FP>, WIDTH>,
+    ) -> Self {
+        let packed_initial_external_constants = external_constants
+            .get_initial_constants()
+            .iter()
+            .map(|arr| arr.map(|c| convert_to_vec_neg_form_neon::<FP>(c.value as i32)))
+            .collect();
+        let packed_terminal_external_constants = external_constants
+            .get_terminal_constants()
+            .iter()
+            .map(|arr| arr.map(|c| convert_to_vec_neg_form_neon::<FP>(c.value as i32)))
+            .collect();
+        Self {
+            external_constants,
+            packed_initial_external_constants,
+            packed_terminal_external_constants,
+        }
```

Comment:
For now at least, we don't need to do any preprocessing for constants as we don't have a custom add_rc_and_sbox function.
```suggestion
/// Represents the external layers (initial and terminal) of the Poseidon2 permutation for NEON-accelerated operations.
#[derive(Debug, Clone)]
pub struct Poseidon2ExternalLayerMonty31<MP: MontyParameters, const WIDTH: usize> {
    /// The original, scalar round constants for both initial and terminal external rounds.
    pub(crate) external_constants: ExternalLayerConstants<MontyField31<MP>, WIDTH>,
}

impl<FP: FieldParameters, const WIDTH: usize> ExternalLayerConstructor<MontyField31<FP>, WIDTH>
    for Poseidon2ExternalLayerMonty31<FP, WIDTH>
{
    fn new_from_constants(
        external_constants: ExternalLayerConstants<MontyField31<FP>, WIDTH>,
    ) -> Self {
        Self {
            external_constants,
        }
```